kratos/python_scripts/restart_utility.py

```diff
@@ -175,9 +187,43 @@ def CreateOutputFolder(self):
                 os.makedirs(folder_path)
             self.model_part.GetCommunicator().GetDataCommunicator().Barrier()
 
-
     #### Protected functions ####
 
+    def _GetRestartFiles(self):
+        """Return a dictionary of stepID - restart_file_list dictionary that stores sets of restart files for each step."""
+        restart_path    = Path(self.__GetFolderPathSave())
+        restart_files   = {}
+        if restart_path.is_dir():
+
+            file_name_data_collector = KratosMultiphysics.FileNameDataCollector(self.model_part, os.path.join(self.__GetFolderPathSave(), self._GetFileNamePattern()), {})
+
+            for file_name_data in file_name_data_collector.GetFileNameDataList():
+                # Get step id
+                if self.restart_control_type_is_time:
+                    step_id = file_name_data.GetTime()
+                else:
+                    step_id = file_name_data.GetStep()
+
+                restart_files[step_id] = file_name_data.GetFileName()
+
+            # barrier is necessary to avoid having some ranks deleting files while other ranks still detect them in the same directory
+            self.model_part.GetCommunicator().GetDataCommunicator().Barrier()
+
+        return restart_files
+
+    def _ClearObsoleteRestartFiles(self):
+        """Delete restart files that are no longer needed."""
+        if self.max_files_to_keep > -1:
+            self.restart_files = self._GetRestartFiles()
+
+            number_of_obsolete_files = len(self.restart_files) - self.max_files_to_keep
+            restart_file_keys = sorted(self.restart_files)
+
```

Comment:
@sunethwarna when adding a print of the names here this is the result I get, seems that the files are not corresponding to the correct rank

```
# rank 0
{0.004: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_0_0.004.rest', 0.008: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_3_0.008.rest', 0.01: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_2_0.01.rest', 0.012: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_0_0.012.rest'}
# rank 1
{0.004: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_0_0.004.rest', 0.008: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_3_0.008.rest', 0.01: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_2_0.01.rest', 0.012: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_0_0.012.rest'}
# rank 2
{0.004: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_0_0.004.rest', 0.008: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_3_0.008.rest', 0.01: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_2_0.01.rest', 0.012: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_0_0.012.rest'}
# rank 3
{0.004: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_0_0.004.rest', 0.008: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_3_0.008.rest', 0.01: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_2_0.01.rest', 0.012: '/home/mate/Repositories/DeckFlutterLB3Medium/Barc.gid/restart/fm/pitch/ured2.0/FluidModelPart_0_0.012.rest'}
```

Is it an issue with the syntax used with restart or does this come from the utility?

Reply:
Not sure if I understand this correctly; are you printing the contents of ```self.restart_files``` in each rank?

The fact that all ranks have the same list of files is intended. It definitely means that multiple ranks will try to delete the same file, but every rank managing its own files wouldn't work if the number of ranks in the old and restarted processes differed.

What's weird is that every step has only one file, and that the ranks of the files across steps aren't identical. Can you print the contents of the restart directory when restarting (ideally before ```RestartUtility``` touches anything)?

Reply:
Oh nevermind, I see the file collection was changed in 68ccdfd.